rangeserver/src/prefetching_buffer.rs

```diff
@@ -0,0 +1,946 @@
+        if let Some(sender) = key_state_sender.get(&key) {
+            let _ = sender.send(KeyState::Requested(fetch_sequence_number));
+        }
+    }
+
+    /// Marks a transaction as being complete by updating bookkeeping to remove
+    /// the request from the data structures
+    pub async fn process_transaction_complete(&self, transaction_id: Uuid) {
+        let mut transaction_keys = self.transaction_keys.lock().await;
+        let mut key_transactions = self.key_transactions.lock().await;
+
+        // Loop over set of transaction keys to call remove key transactions (accept only transaction id)
+        if let Some(keys) = transaction_keys.get(&transaction_id) {
+            for key in keys {
+                // Log that this key is no longer being requested by this transaction
+                self.remove_from_key_transactions(
+                    &mut key_transactions,
+                    transaction_id,
+                    key.clone(),
+                );
+                // If the key is no longer requested by any transaction, also delete it from the BTree
+                if !key_transactions.contains_key(key) {
+                    self.delete(key.clone()).await;
+                }
+            }
+        }
+        // Remove transaction from transaction_key
+        let _ = transaction_keys.remove(&transaction_id);
+    }
+
+    /// Update the prefetch_store and update the key_state
+    pub async fn upsert(&self, key: Bytes, value: Bytes) {
+        let mut prefetch_store = self.prefetch_store.lock().await;
+        let mut key_state = self.key_state.lock().await;
+        // If key is still being requested by a different transaction, update BTree with latest value
+        if self
+            .key_transactions
+            .lock()
+            .await
+            .contains_key(&key.clone())
+        {
+            let _ = prefetch_store.insert(key.clone(), value);
+            // If the key is still loading in a prefetch, change to fetched
+            if let KeyState::Loading(_) = key_state.get(&key).unwrap() {
+                key_state.insert(key.clone(), KeyState::Fetched);
+                // Notify all watchers of the state change
+                if let Some(sender) = self.key_state_sender.lock().await.get(&key) {
+                    let _ = sender.send(KeyState::Fetched);
+                }
+            }
+        } else {
+            // If key is not being requested by any transactions, remove from prefetch_store instead of update
+            let _ = prefetch_store.remove(&key); // might be None
+            let _ = key_state.remove(&key); // might be None
+        }
+    }
+
+    /// If a transaction deletes a key, it must also be deleted from the buffer
+    /// This function deletes a key from prefetch_store and key_state
+    pub async fn delete(&self, key: Bytes) {
+        let mut prefetch_store = self.prefetch_store.lock().await;
+        let mut key_state = self.key_state.lock().await;
```

Comment:
we should not remove the key_state for a deleted key! we want to maintain in memory the fact that this key does not exist, so that a transaction does not need to go to storage to learn that fact.

Reply:
I added a KeyState::Deleted to maintain this state. We should discuss the key_state logic if a value for a key is reinserted after deletion.